src/stream.rs

```diff
@@ -0,0 +1,216 @@
+        }
+        next_item
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::StreamBackoff;
+    use crate::{backoff::Backoff, future::TokioSleeper};
+    use futures_channel::mpsc;
+    use futures_util::{pin_mut, poll, stream, StreamExt};
+    use std::{task::Poll, time::Duration};
+    use tokio_1 as tokio;
+
+    #[tokio::test]
+    async fn stream_should_back_off() {
+        tokio::time::pause();
+        let tick = Duration::from_secs(1);
+        let rx = stream::iter([Ok(0), Ok(1), Err(2), Ok(3), Ok(4)]);
+        let rx = StreamBackoff::new(rx, crate::backoff::Constant::new(tick), TokioSleeper);
+        pin_mut!(rx);
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(0))));
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(1))));
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Err(2))));
+        assert_eq!(poll!(rx.next()), Poll::Pending);
+        tokio::time::advance(tick * 2).await;
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(3))));
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(4))));
+        assert_eq!(poll!(rx.next()), Poll::Ready(None));
+    }
+
+    #[tokio::test]
+    async fn backoff_time_should_update() {
+        tokio::time::pause();
+        let (tx, rx) = mpsc::unbounded();
+        let rx = StreamBackoff::new(rx, LinearBackoff::new(Duration::from_secs(2)), TokioSleeper);
+        pin_mut!(rx);
+        tx.unbounded_send(Ok(0)).unwrap();
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(0))));
+        tx.unbounded_send(Ok(1)).unwrap();
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(1))));
+        tx.unbounded_send(Err(2)).unwrap();
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Err(2))));
+        assert_eq!(poll!(rx.next()), Poll::Pending);
+        tokio::time::advance(Duration::from_secs(3)).await;
+        assert_eq!(poll!(rx.next()), Poll::Pending);
+        tx.unbounded_send(Err(3)).unwrap();
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Err(3))));
+        tx.unbounded_send(Ok(4)).unwrap();
+        assert_eq!(poll!(rx.next()), Poll::Pending);
+        tokio::time::advance(Duration::from_secs(3)).await;
+        assert_eq!(poll!(rx.next()), Poll::Pending);
+        tokio::time::advance(Duration::from_secs(2)).await;
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(4))));
+        assert_eq!(poll!(rx.next()), Poll::Pending);
+        drop(tx);
+        assert_eq!(poll!(rx.next()), Poll::Ready(None));
+    }
+
+    #[tokio::test]
+    async fn backoff_should_close_when_requested() {
+        assert_eq!(
+            StreamBackoff::new(
+                stream::iter([Ok(0), Ok(1), Err(2), Ok(3)]),
+                crate::backoff::Stop {},
+                TokioSleeper
+            )
+            .collect::<Vec<_>>()
+            .await,
+            vec![Ok(0), Ok(1), Err(2)]
+        );
+    }
+
+    /// Dynamic backoff policy that is still deterministic and testable
+    struct LinearBackoff {
```

Comment:
Might be worth promoting this to `crate::backoff`? Not sure I'd use it in production, but it's pretty nice to have a predictable and deterministic variant of `ExponentialBackoff` when writing marble tests.